Fix issues with the search

ryanlerch opened the pull request:
This PR contains 4 commits. 

First, updates the typeahead.js library to the latest upstream version 0.11.1. We need the newer version of this library for the 3 fixes in this PR. The typeahead API changed between the 0.10.x release we were using before, so this commit also updates our code to work with the new API.

Secondly, a spinner now shows in the search-results dropdown when we are fetching results from the bodhi server. Each of the 4 results sections has their own spinner. Fixes #436

Thirdly, middle clicking on search results now opens the result in a new tab as expected. This was acheived by wrapping all search results in an a element. Fixes #461

Finally, we escape the packagename that we pass as a URI for cases like memtest86+ where the package name includes a special character that should not be in the URI. Fixes #1015

ryanlerch committed:
Update Typeahead JS library to version 1.1.1

Updates the typeahead bundle (containing both bloodhound and
typeahead) to the corejavascript maintained fork version 1.1.1.

https://github.com/corejavascript/typeahead.js/releases/tag/v1.1.1

The maintained fork was used instead of the unmaintained upstream
as we were hitting an issue where on some occasions not all the
search results were shown.

The typeahead API changed between releases, so this also makes
our search work with the new API.

ryanlerch committed:
Add spinner on search while async calls in progress

Adds spinners on the search results dropdown when searching,
so the user knows if the search is still underway. There is a
seperate spinner for each of the 4 categories in the search results

Fixes: #436

ryanlerch committed:
make middle click on search results work as expected

Adds an additional <a> element on search results in the
typeahead dropdown so middle-clicking on search results
opens the result in a new tab as expected.

fixes: #461

ryanlerch committed:
encode the URIs of packagename search results

encodes the URI for the result of the packagename search, as it
may include special charaters like + that we dont want in the URL.

Fixes: #1015

bowlofeggs commented:
jenkies test

bowlofeggs reviewed (changes requested):
Please write a release note mentioning the new version of the typeahead library so I remember to update the spec file to declare that new version.

I did notice that searching for "memtest86+" did fix the package search, but the update search still shows no results. Additionally, even "memtest8" doesn't find update results, but "memtest" does find the memtest86+ results. On production, "memtest8" does find the updates, so I think this might be a small regression.

ryanlerch commented:
@bowlofeggs after a lot of digging, finally figured out that your issue with 'memtest8' as a search was actually a bug in 0.11.1 of typeahead.js. However, this is where this gets messy, it appears that the main typeahead.js (https://github.com/twitter/typeahead.js) is unmaintained. There are a list of bugs on this issue (https://github.com/twitter/typeahead.js/issues/1381) of people having similar issues. 

There is however, a maintained fork of typeahead,js that contains the fix for the issue we were having: https://github.com/corejavascript/typeahead.js . In this PR i am now using their 1.1.1 release which solves the issue.

ryanlerch commented:
Also have updated the release notes with a note about the updated typeahead.js library.

bowlofeggs commented:
jenkies test

bowlofeggs reviewed (dismissed):
Cool, this looks good to merge!

I did notice one remaining issue, but it's not introduced by this change (so it can be fixed separately if you prefer): If you type "memtest86+" and hit enter (without scrolling down to the search results), it still takes you to a URL with an unescaped "+" in it, which doesn't work. Again, this is how it is in production too, so we can fix it later if you prefer. Feel free to merge this as you please!

ryanlerch committed:
encode value when searching by pressing enter

the search.js script has a special method that catches
if the user presses enter on the search bar without choosing
an item from the typeahead dropdown, and just searches for the
package name. WE previously escaped the search value for items
chosen in the dropdown, but missed the special case when the user
presses enter. This escapes that value, so results for packagenames
with special URI chars in them (e.g. memtest86+) now show properly
when just typing a search string and pressing enter.

ryanlerch commented:
@bowlofeggs thanks for catching this one. Just added an extra commit to this PR to fix this. 

the search.js script has a special method that catches if the user presses enter on the search bar without choosing an item from the typeahead dropdown, and just searches for the package name. WE previously escaped the search value for items chosen in the dropdown, but missed the special case when the user presses enter. This escapes that value, so results for packagenames with special URI chars in them (e.g. memtest86+) now show properly when just typing a search string and pressing enter.

bowlofeggs reviewed (approved):
Looks great, thanks @ryanlerch!
